feat(settings): add preferredOrg field to folder configuration

bastiandoetsch opened the pull request:
### Description

This PR adds support for configuring a preferred organization at the project level through the FolderConfig settings, and deprecates the global organization field in favor of this new project-level setting.

**Changes:**
- Added `preferredOrg` field to `FolderConfig` data class with default empty string value
- Added UI input field for preferred organization in SnykSettingsDialog, positioned below "Additional Parameters" in Project settings
- Implemented `getPreferredOrg()` method in FolderConfigSettings to retrieve the project's preferred organization
- Updated SnykProjectSettingsConfigurable to properly save and load preferredOrg settings
- **Deprecated global organization field** - Disabled the global organization text field in General settings and updated tooltip to guide users to use the project-level setting instead
- Added comprehensive test coverage with 10 new tests for preferredOrg functionality
- Fixed compilation errors in existing tests due to updated FolderConfig constructor

**Jira:** IDE-1366

The preferredOrg field allows users to configure a preferred organization at the project level, with the same behavior as Additional Parameters. The global organization setting in General settings is now deprecated in favor of this project-level configuration.

### Checklist

- [x] Read and understood the [Code of Conduct](https://github.com/snyk/snyk-intellij-plugin/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/snyk/snyk-intellij-plugin/blob/master/CONTRIBUTING.md).
- [x] Tests added and all succeed
- [x] Linted
- [ ] CHANGELOG.md updated
- [ ] README.md updated, if user-facing

### Screenshots / GIFs

_UI changes:_
- New "Preferred organization" field added in Project settings section, positioned below "Additional Parameters"
<img width="804" height="120" alt="image" src="https://github.com/user-attachments/assets/ff4788c4-60cb-4a78-9912-c41ac8b3a591" />
- Global "Organization" field in General settings is now disabled with a deprecation notice in the tooltip
<img width="975" height="71" alt="image" src="https://github.com/user-attachments/assets/c9f3dac8-6d52-4b8d-9c2a-09c7d1d39529" />

bastiandoetsch committed:
feat(settings): add preferredOrg field to folder configuration

- Add preferredOrg field to FolderConfig with default empty string value
- Add preferredOrg input field in SnykSettingsDialog below Additional Parameters
- Add getPreferredOrg method to FolderConfigSettings to retrieve project's preferred org
- Update SnykProjectSettingsConfigurable to save/load preferredOrg settings
- Add comprehensive test coverage with 10 new tests for preferredOrg functionality
- Fix compilation errors in existing tests with updated FolderConfig constructor

The preferredOrg field allows users to configure a preferred organization
at the project level, with the same behavior as Additional Parameters.

bastiandoetsch commented:
Failing test is due to missing CLI with protocol version 21. This is ok and can be ignored.

bastiandoetsch committed:
fix: check for disposal when getting message bus publisher

bastiandoetsch committed:
fix: global org field deactivation and tooltip

bastiandoetsch committed:
feat: add checkbox to settings for org autoselect

bastiandoetsch committed:
chore: update layout for auto-select checkbox

acke committed:
fix: handle checkbox for autoorg

fix: handle windows uris correct in test

acke committed:
fix: update the auto org checkbox and field to handle correctly

acke committed:
feat: update organization help text and add ContextHelpLabel [IDE-1366]

- Update auto-org help text to be more concise and clear
- Add ContextHelpLabel for preferred organization field with help text
- Add link to user docs for auto-org feature
- Remove tooltip from preferredOrgTextField in favor of ContextHelpLabel
- Create preferredOrgPanel to contain text field and help label

acke committed:
feat: re-enable global organization support with fallback logic [IDE-1366]

- Update FolderConfigSettings.getOrganization() to fallback to global organization when folder-specific values are empty
- Update FolderConfigSettings.getPreferredOrg() to fallback to global organization when folder-specific preferred org is empty
- Update UI initialization to show global organization as fallback in preferredOrgTextField
- Update updatePreferredOrgTextField() to use global organization as fallback
- Add null safety checks for ApplicationManager in test environments
- Maintain backward compatibility with existing folder-specific settings

acke committed:
feat: always think sast is enabled [IDE-1531] [IDE-1532] (#738)

* fix: removed sast and local code engine from UI components

* fix: remove unused functions and cleanup

andrewrobinsonhodges-snyk committed:
fix: alignment and wording of help text for orgnaization settings

andrewrobinsonhodges-snyk committed:
refactor: use dimension instead of deprecated width and height

andrewrobinsonhodges-snyk committed:
Merge branch 'master' into feat/IDE-1366_get-project-org-from-ls

andrewrobinsonhodges-snyk committed:
fix: preferred org dialog box should be enabled based on auto determination settings

andrewrobinsonhodges-snyk committed:
fix: add message bus for updating folder configs

andrewrobinsonhodges-snyk committed:
fix: typo in org selection help text strings

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/snyk/snyk-intellij-plugin?pullRequest=735) <br/>All committers have signed the CLA.

andrewrobinsonhodges-snyk committed:
refactor: rename org selection variables for clarity, and update strings to align with other IDEs

andrewrobinsonhodges-snyk committed:
fix: update link text for organization

acke committed:
fix: support empty preferred org, fallback to global org

nick-y-snyk committed:
fix: display Snyk Code disabled in the tree view [IDE 1538] (#741)

* fix: race condition for showing error in the tree view, remove notification when Snyk Code isn't enabled for folder

* docs: move manual test plan to docs directory [IDE-1507]

- Move IDE-1366_manual_test_plan.md from root to docs/IDE-1507_manual-test-plan.md
- Organize test documentation in docs directory

* feat: migrate to PresentableError

---------

Co-authored-by: Knut Funkel <knut.funkel@snyk.io>

andrewrobinsonhodges-snyk committed:
Merge branch 'master' into feat/IDE-1366_get-project-org-from-ls

acke committed:
fix: removed the manual test plan

rrama commented on src/main/kotlin/snyk/common/lsp/Types.kt:
```diff
@@ -558,11 +573,14 @@ data class FolderConfigsParam(
     @SerializedName("referenceFolderPath") val referenceFolderPath: String? = "",
     @SerializedName("scanCommandConfig") val scanCommandConfig: Map<String, ScanCommandConfig>? = emptyMap(),
+    @SerializedName("orgSetByUser") val orgSetByUser: Boolean = false,
 ) : Comparable<FolderConfig> {
```
Do we need to capture and send back the SAST settings and feature flags?

andrewrobinsonhodges-snyk committed:
fix: revert change to settings dialog behaviour dependent on org modification

andrewrobinsonhodges-snyk committed:
refactor: remove unneeded dialog box listener

rrama commented on src/main/kotlin/snyk/common/lsp/settings/FolderConfigSettings.kt:
```diff
@@ -82,4 +88,128 @@ class FolderConfigSettings {
+     * @param project the project to get the organization for
+     * @return the appropriate organization for the project
+     */
+    fun getOrganization(project: Project): String {
```
Unused, delete.

rrama commented on src/main/kotlin/snyk/common/lsp/settings/FolderConfigSettings.kt:
```diff
@@ -82,4 +88,128 @@ class FolderConfigSettings {
+            .asSequence()
+            .filter { languageServerWrapper.configuredWorkspaceFolders.contains(it) }
+            .map { getFolderConfig(it.uri.fromUriToPath().toString()) }
+            .firstOrNull()
```
Nit pick: Can this be extracted as a util function, my brain hurts reading Kotlin like this and this is the type of place a typo could slip in and I would never notice.

rrama commented on src/main/kotlin/snyk/common/lsp/settings/FolderConfigSettings.kt:
```diff
@@ -82,4 +88,128 @@ class FolderConfigSettings {
+            .asSequence()
+            .filter { languageServerWrapper.configuredWorkspaceFolders.contains(it) }
+            .map { getFolderConfig(it.uri.fromUriToPath().toString()) }
+            .filter { it.preferredOrg.isNotEmpty() }
```
Scrap this filter.

rrama commented on src/test/kotlin/snyk/common/lsp/settings/FolderConfigSettingsTest.kt:
```diff
@@ -303,4 +317,501 @@ class FolderConfigSettingsTest {
+        assertEquals("Should return new organization", "new-org", result)
+
+        // Clear organization
+        settings.setOrganization(projectMock, null)
```
Is setting to `null` valid or should it be `""`?

andrewrobinsonhodges-snyk committed:
refactor: remove unused code

andrewrobinsonhodges-snyk committed:
fix: code review markups

andrewrobinsonhodges-snyk committed:
fix: update org selection help text

andrewrobinsonhodges-snyk committed:
fis: remove unneeded folder config filter

andrewrobinsonhodges-snyk committed:
test: remove unneeded tests

rrama reviewed (approved):
Get it shipped!
